src/tls/openssl/tls.c

```diff
@@ -1423,3 +1445,329 @@ void tls_disable_verify_server(struct tls *tls)
 
 	tls->verify_server = false;
 }
+
+
+/**
+ * Set minimum TLS version
+ *
+ * @param tls     TLS Object
+ * @param version Minimum version, e.g.: TLS1_2_VERSION
+ */
+int tls_set_min_proto_version(struct tls *tls, int version)
+{
+	if (!tls)
+		return EINVAL;
+
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L)
+	if (SSL_CTX_set_min_proto_version(tls->ctx, version))
+		return 0;
+	else
+#endif
+		return EACCES;
+
+}
+
+
+/**
+ * Set maximum TLS version
+ *
+ * @param tls     TLS Object
+ * @param version Maximum version, e.g. TLS1_2_VERSION
+ */
+int tls_set_max_proto_version(struct tls *tls, int version)
+{
+	if (!tls)
+		return EINVAL;
+
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L)
+	if (SSL_CTX_set_max_proto_version(tls->ctx, version))
+		return 0;
+	else
+#endif
+		return EACCES;
+}
+
+
+struct session_entry {
+	struct le le;
+	struct sa peer;
+	SSL_SESSION *sess;
+};
+
+
+static void session_destructor(void *arg)
+{
+	struct session_entry *e = arg;
+
+	hash_unlink(&e->le);
+	if (e->sess)
+		SSL_SESSION_free(e->sess);
+}
+
+
+static bool session_cmp_handler(struct le *le, void *arg)
+{
+	const struct session_entry *s = le->data;
+	if (!s)
+		return false;
+
+	return sa_cmp(&s->peer, arg, SA_ALL);
+}
+
+
+static int tls_session_update_cache(const struct tls_conn *tc,
+	SSL_SESSION *sess)
+{
+	struct sa peer;
+	struct session_entry* e = NULL;
+	int err = 0;
+
+	if (!tc || !tc->tls) {
+		DEBUG_WARNING("%s: no tc or tls.\n", __func__);
+		return EINVAL;
+	}
+
+	err = tcp_conn_peer_get(tls_get_tcp_conn(tc), &peer);
+	if (err) {
+		DEBUG_WARNING("%s: tcp_conn_peer_get failed: (%m).\n",
+			__func__, err);
+		return ENODATA;
+	}
+
+	e = list_ledata(hash_lookup(tc->tls->reuse.ht_sessions,
+					     sa_hash(&peer, SA_ALL),
+					     session_cmp_handler, &peer));
+	e = mem_deref(e);
```

Comment:
src/tls/openssl/tls.c:1540:2: warning: Value stored to 'e' is never read [deadcode.DeadStores]
        e = mem_deref(e);
        ^   ~~~~~~~~~~~~
1 warning generated.

Reply:
`mem_deref(e);` Should be fine.